accounts/accounts_cli.go

```diff
@@ -0,0 +1,328 @@
+
+	urlPath := a.host + fmt.Sprintf("/query/upload/lastentry/%s", acc.Id)
+
+	req, _ := http.NewRequest("GET", urlPath, nil)
+	req.Header.Add("x-tidepool-session-token", a.User.Token)
+
+	res, err := a.client.Do(req)
+
+	if err != nil {
+		return errors.New("Could attempt to find the last upload: " + err.Error())
+	}
+
+	switch res.StatusCode {
+	case 200:
+		data, err := ioutil.ReadAll(res.Body)
+
+		if err != nil {
+			log.Println("Error trying to read the last upload data", err.Error())
+			return nil
+		}
+
+		json.Unmarshal(data, &acc.LastUpload)
+		return nil
+	default:
+		log.Printf("Failed finding last upload info [%d] for [%s]", res.StatusCode, acc.Profile.FullName)
+		return nil
+	}
+}
+
+func (a *admin) login(un, pw string) error {
+
+	urlPath := a.host + "/auth/login"
+
+	req, _ := http.NewRequest("POST", urlPath, nil)
+	req.SetBasicAuth(un, pw)
+
+	res, err := a.client.Do(req)
+	if err != nil {
+		return errors.New(fmt.Sprint("Login request failed", err.Error()))
+	}
+
+	switch res.StatusCode {
+	case 200:
+		data, _ := ioutil.ReadAll(res.Body)
+		json.Unmarshal(data, &a.User)
+		a.User.Token = res.Header.Get("x-tidepool-session-token")
+		return nil
+	default:
+		return errors.New(fmt.Sprint("Login failed", res.StatusCode))
+	}
+}
+
+func (a *admin) findAccounts() error {
+
+	urlPath := a.host + fmt.Sprintf("/access/groups/%s", a.User.Id)
+
+	req, _ := http.NewRequest("GET", urlPath, nil)
+	req.Header.Add("x-tidepool-session-token", a.User.Token)
+
+	res, err := a.client.Do(req)
+
+	if err != nil {
+		return errors.New("Could attempt to find administered accounts: " + err.Error())
+	}
+
+	switch res.StatusCode {
+	case 200:
+		data, _ := ioutil.ReadAll(res.Body)
+
+		var raw map[string]interface{}
+
+		json.Unmarshal(data, &raw)
+
+		for key, value := range raw {
+			a.Accounts = append(a.Accounts, &Account{Id: string(key), Perms: value})
+		}
+		return nil
+	default:
+		log.Println("Failed finding profiles", res.StatusCode)
+		return nil
+	}
+}
+
+func (a *admin) loadExistingReport(path string) error {
+	log.Println("loading audit report ...")
+
+	rf, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	jsonParser := json.NewDecoder(rf)
+	if err = jsonParser.Decode(&a); err != nil {
+		return err
+	}
+	rf.Close()
+	return nil
+}
+
+func (a *admin) accountDetails() {
+	var wg sync.WaitGroup
+
+	for _, account := range a.Accounts {
+		// Increment the WaitGroup counter.
+		wg.Add(1)
+		// Launch a goroutine to account data
+		go func(account *Account) {
+			// Decrement the counter when the goroutine completes.
+			defer wg.Done()
+			// Fetch the account data
+			a.findPublicInfo(account)
+		}(account)
+	}
+	// Wait for all fetches to complete.
+	wg.Wait()
+	return
+}
+func (a *admin) audit() {
+
+	var wg sync.WaitGroup
+
+	for _, account := range a.Accounts {
+		// Increment the WaitGroup counter.
+		wg.Add(1)
+		// Launch a goroutine to account data
+		go func(account *Account) {
```

Comment:
ahh yeah - that wasn't my intention. It will "work as coded" (i.e. kill things ) but my intention is to wait for each call so finish so that I didn't flood the system :)